contracts/campaignAirdrops/ExistingERC20DropCampaign.sol

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.14;
+
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "../interfaces/AirdropMerkleProof.sol";
+import "../interfaces/IAirBroFactory.sol";
+
+/// @title Airdrops existing ERC20 tokens for airdrop recipients
+contract ExistingERC20DropCampaign is AirdropMerkleProof {
+    using SafeERC20 for IERC20;
+
+    IERC20 public immutable rewardToken;
+    address public immutable airbroCampaignFactoryAddress;
+    uint256 public immutable tokenSupply;
+
+    string public airdropType = "ERC20";
+    bool public airdropFunded;
+    bool public merkleRootSet;
+    uint256 internal numberOfClaimers;
+    uint256 public tokensPerClaim;
+    uint256 public airdropExpirationTimestamp;
+    bytes32 public merkleRoot;
+
+    mapping(address => bool) public hasClaimed;
+
+    address internal airdropFunder;
+
+    event MerkleRootChanged(bytes32 merkleRoot, uint256 _numberOfClaimers);
+    event Claimed(address indexed claimer);
+    event AirdropFunded(address contractAddress);
+
+    error AlreadyFunded();
+    error Unauthorized();
+    error AlreadyRedeemed();
+    error AirdropStillActive();
+    error AirdropExpired();
+    error NotEligible();
+    error InsufficientAmount();
+    error InsufficientLiquidity();
+    error MerkleRootAlreadySet();
+
+    modifier onlyAdmin() {
+        if (msg.sender != IAirBroFactory(airbroCampaignFactoryAddress).admin()) revert Unauthorized();
+        _;
+    }
+
+    constructor(
+        address _rewardToken,
+        uint256 _tokenSupply,
+        address _airbroCampaignFactoryAddress
+    ) {
+        rewardToken = IERC20(_rewardToken);
+        tokenSupply = _tokenSupply;
+        airbroCampaignFactoryAddress = _airbroCampaignFactoryAddress;
+    }
+
+    /// @notice Sets the merkleRoot and the number of claimers (also setting the amount each claimer receivers)
+    /// @notice can only be done by admin
+    /// @param _merkleRoot The root hash of the Merle Tree
+    /// @param _numberOfClaimers The number of people eligible to claim
+    function setMerkleRoot(bytes32 _merkleRoot, uint256 _numberOfClaimers) external onlyAdmin {
+        if (merkleRootSet) revert MerkleRootAlreadySet();
+
+        merkleRootSet = true;
+        merkleRoot = _merkleRoot;
+        airdropExpirationTimestamp = block.timestamp + 60 days;
```

Comment:
Maybe constructor is a better place for this + it be adjusted and immutable to save some gas

Reply:
What are you referring to specifically?

Reply:
I kinda see it that Admin will deploy this contract, and then setMerkleRoot the same day / or shortly after. It would be cheaper to initialize this `airdropExpirationTimestamp` in constructor like this `airdropExpirationTimestamp = block.timestamp + 60 days;`
OR
have it as construcotr param `_airdropExpirationTimestamp`
and set it when deploying to whatever you want, 5 days, 30 days, 60 days 
That way it looks more flexibile to me, maybe somebody doesn't want 60 days expiration time?

Reply:
So basically the merkleRoot is set when the backend decides the campaign is finished, it might happen 1 month after creating the campaign. 

After setting it, all users who have a valid merkleProof have 60 days to claim their reward. 

If they do not claim this reward within 60 days, the funder of the contract (might be the admin but also might not be), can withdraw any remaining funds from the contract, this is to avoid permanently locking funds if someone doesn't withdraw the rewad.